app/models/user.rb

```diff
@@ -351,6 +351,32 @@ def update_firecloud_status
     end
   end
 
+  # determine if user needs to accept an updated Terra Terms of Service agreement
+  def must_accept_terra_tos?
```

Comment:
Consider changing this method to just be 
 `has_accepted_latest_terra_tos?` and then having the callbacks_controller be the place where we have the logic for not calling the method if they aren't registered.  That lets this method focus purely on the terra ToS logic, and the controller owns the 'when do we prompt' logic

Reply:
TL;DR - you're right, but because of weirdness in how we emit/handle errors from `FireCloudClient`, not making the registration check in the User instance method leads to a corner case where a non-registered user would have `must_accept_terra_tos?` return `true`, rather than `false`.  Because we only emit `RuntimeError` from orchestration API requests, we don't know if the real HTTP response code was `401`, `403` or `404`.  The controller check sidesteps this, but it leaves me with a bad feeling knowing that we have a method that has a known issue that external logic avoids.  Hence the pre-flight check for registration status.

Ideally, we'd do a larger refactoring to emit correct errors from `FireCloudClient`, like `HcaAzulClient` and `DataRepoClient` do, because then we could catch the `404` and handle appropriately.